lang/src/api.rs

```diff
@@ -69,34 +69,72 @@ pub enum TypeDescription {
     Address,
     /// The SRML balance type.
     Balance,
+    /// The tuple type
+    Tuple {
+        elems: Vec<TypeDescription>,
+    },
+    /// The fixed size array type
+    Array {
+        inner: Box<TypeDescription>,
+        arity: u32,
+    }
 }
 
 impl TryFrom<&syn::Type> for TypeDescription {
     type Error = Errors;
 
     fn try_from(ty: &syn::Type) -> Result<Self> {
         use quote::ToTokens;
-        match ty.into_token_stream().to_string().as_str() {
-            "bool" => Ok(TypeDescription::Bool),
-            "u8" => Ok(TypeDescription::U8),
-            "u16" => Ok(TypeDescription::U16),
-            "u32" => Ok(TypeDescription::U32),
-            "u64" => Ok(TypeDescription::U64),
-            "u128" => Ok(TypeDescription::U128),
-            "i8" => Ok(TypeDescription::I8),
-            "i16" => Ok(TypeDescription::I16),
-            "i32" => Ok(TypeDescription::I32),
-            "i64" => Ok(TypeDescription::I64),
-            "i128" => Ok(TypeDescription::I128),
-            "Address" => Ok(TypeDescription::Address),
-            "Balance" => Ok(TypeDescription::Balance),
-            unsupported => {
-                bail!(
-                    ty,
-                    "{} is unsupported as message interface type",
-                    unsupported
-                )
+        let primitive = |ty: &syn::Type| {
+            match ty.into_token_stream().to_string().as_str() {
+                "bool" => Ok(TypeDescription::Bool),
+                "u8" => Ok(TypeDescription::U8),
+                "u16" => Ok(TypeDescription::U16),
+                "u32" => Ok(TypeDescription::U32),
+                "u64" => Ok(TypeDescription::U64),
+                "u128" => Ok(TypeDescription::U128),
+                "i8" => Ok(TypeDescription::I8),
+                "i16" => Ok(TypeDescription::I16),
+                "i32" => Ok(TypeDescription::I32),
+                "i64" => Ok(TypeDescription::I64),
+                "i128" => Ok(TypeDescription::I128),
+                "Address" => Ok(TypeDescription::Address),
+                "Balance" => Ok(TypeDescription::Balance),
+                unsupported => {
+                    bail!(
+                        ty,
+                        "{} is unsupported as message interface type",
+                        unsupported
+                    )
+                }
+            }
+        };
+        match ty {
+            syn::Type::Tuple(tuple) => {
+                let elems = tuple
+                    .elems
+                    .iter()
+                    .map(primitive)
```

Comment:
Interesting approach with the primitives map. However, as far as I can see this restricts us from having recursive types like `(u32, (bool, u32))` or `[[u32; 5]; 5]`. I think this is a simple fix by again moving the primitives into the root match and together with my suggestion above just simply call `TypeDescription::try_from` recursively in the `TupleTypeDescription` and `ArrayTypeDescription` respectively.